ADAL/src/cache/ADResponseCacheHandler.m

```diff
@@ -88,7 +88,8 @@ + (ADAuthenticationResult *)handleError:(NSError *)msidError
                                   cache:(MSIDLegacyTokenCacheAccessor *)cache
                                  params:(ADRequestParameters *)requestParams
 {
-    if (response.oauthErrorCode == MSIDErrorServerInvalidGrant && refreshToken)
+    NSString *subError = [[msidError userInfo] objectForKey:MSIDOAuthSubErrorKey];
+    if (response.oauthErrorCode == MSIDErrorServerInvalidGrant && refreshToken && (subError == nil || [subError caseInsensitiveCompare:@"consent_required"] != NSOrderedSame))
```

Comment:
Why are we not removing tokens for any invalid grant?

Reply:
consent_required does not imply that the refresh token is invalid for all resources, only that interactive flow is required.

Reply:
Can we please update this to following?

`if (response.oauthErrorCode == MSIDErrorServerInvalidGrant && refreshToken && [subError caseInsensitiveCompare:@"consent_required"] != NSOrderedSame)`

That way we only exclude `consent_required` for the time being as agreed. Your condition will also stop removing tokens for valid conditions when token should be removed. Thanks!

Reply:
I believe the logic is correct as is, and removes the token for any MSIDErrorServerInvalidGrant error other than consent_required. The difference between these two conditions is that I am considering the case where subError is nil. In this case nil == 0 and NSOrderedSame == 0 so it will not remove the token. I just did extra testing around this condition to confirm this is the case.